src/bcf/record.rs

```diff
@@ -79,6 +79,30 @@ impl NumericUtils for i32 {
     }
 }
 
+/// A trait to allow for seamless use of bytes or integer identifiers for filters
+pub trait FilterId {
+    fn id_from_header(&self, header: &HeaderView) -> Result<Id>;
```

Comment:
As none of the function implementations can return an error, I don't think wrapping this in a `Result` is needed:
```suggestion
    fn id_from_header(&self, header: &HeaderView) -> Id;
```

Maybe this is what is causing the overhead for `Id`? Because we only need the unpacking after using `name_to_id()`, so maybe move the unpacking in to the implementation for `[u8]`?